### PR TITLE
 TIG-1337 add remaining genny canaries

### PR DESCRIPTION
**Pull request by guoyr**
Add genny canaries for CPU, L2 cache, L3 cache, and sleep/context switching.

### CODE REVIEW

File: src/canaries/include/canaries/Workloads.hpp [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+#include <bsoncxx/builder/basic/document.hpp>
+#include <bsoncxx/builder/basic/kvp.hpp>
+
+#include <mongocxx/client.hpp>
+
+#include <gennylib/v1/ConfigNode.hpp>
+#include <gennylib/v1/PoolManager.hpp>
+
+namespace genny::canaries {
```
Comment (guoyr) [inline src/canaries/include/canaries/Workloads.hpp]: The tasks in this file are ported over from https://github.com/mongodb/mongo/blob/r4.1.6/src/mongo/unittest/system_resource_canary_bm.cpp

File: src/canaries/include/canaries/Loops.hpp [hunk 2]
```diff
@@ -31,30 +33,30 @@ using Nanosecond = int64_t;
...
 class Loops {
 public:
     explicit Loops(int64_t iterations) : _iterations(iterations){};
 
     /**
-     * Run a basic for-loop, used as the baseline.
+     * Run native for-loop; used as the control group with no Genny code.
      */
-    Nanosecond simpleLoop();
+    Nanosecond simpleLoop(Args&&... args);
```
Comment (rtimmons) [inline src/canaries/include/canaries/Loops.hpp]: can you document here or in the class what the `Args` are for?

File: src/canaries/include/canaries/Workloads.hpp [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    int64_t _inc() {
+        // Do a trivial amount of work to (hopefully) trick the compiler
+        // into not optimizing everything away.
+        return _counter++;
+    }
+};
+
+class SleepTask : public Task {
+public:
+    inline void run() override {
```
Comment (rtimmons) [inline src/canaries/include/canaries/Workloads.hpp]: 🤷‍♂️ class functions defined in headers are implicitly `inline`

File: src/canaries/include/canaries/Workloads.hpp [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+namespace genny::canaries {
+
+class Task {
+public:
+    virtual inline void run() = 0;
+
+protected:
+    template <class T>
+    inline void doNotOptimize(T const& value) {
+        asm volatile("" : : "r,m"(value) : "memory");
```
Comment (rtimmons) [inline src/canaries/include/canaries/Workloads.hpp]: 🤯 is this..documented somewhere? or copied from somewhere?

File: src/canaries/include/canaries/Workloads.hpp [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+public:
+    L2Task() : _strider(_arrLength) {}
+
+    inline void run() override {
+        char** dummyResult = 0;  // Dummy result to prevent the loop from being optimized out.
+        char** ptrToNextLinkedListNode = reinterpret_cast<char**>(_strider.data.get()[0]);
+
+        for (size_t i = 0; i < _counter; ++i) {
+            HUNDRED;
```
Comment (rtimmons) [inline src/canaries/include/canaries/Workloads.hpp]: fancy!

File: src/canaries/include/canaries/Workloads.hpp (hunk 4 above)
Comment (guoyr) [reply src/canaries/include/canaries/Workloads.hpp]: I added a comment. It's adapted from Google Benchmark's namesake function.

**Comment by guoyr** [comment]
I added a comment with a link to the original version of these fancy workloads in the mongo repo. There're more comments and explanations there that I don't feel like duplicating here. I did add comments on modifications.

**Pull request merged**